lens/registry.go

```diff
@@ -0,0 +1,335 @@
+			r.lensWarehouse[cfg.SourceSchemaVersionID] = locker
+		}
+
+	drainLoop:
+		for {
+			select {
+			case <-locker.safes:
+			default:
+				break drainLoop
+			}
+		}
+
+		for _, lensPipe := range newLensPipes {
+			locker.returnLens(lensPipe)
+		}
+
+		r.lensConfigs[cfg.SourceSchemaVersionID] = cfg
+	})
+
+	return nil
+}
+
+func (r *lensRegistry) RefreshLenses(ctx context.Context, txn datastore.Txn) error {
+	prefix := core.NewSchemaVersionMigrationKey("")
+	q, err := txn.Systemstore().Query(ctx, query.Query{
+		Prefix: prefix.ToString(),
+	})
+	if err != nil {
+		return err
+	}
+
+	for res := range q.Next() {
+		// check for Done on context first
+		select {
+		case <-ctx.Done():
+			// we've been cancelled! ;)
+			err = q.Close()
+			if err != nil {
+				return err
+			}
+
+			return nil
+		default:
+			// noop, just continue on the with the for loop
+		}
+
+		if res.Error != nil {
+			err = q.Close()
+			if err != nil {
+				return err
+			}
+			return res.Error
+		}
+
+		var cfg client.LensConfig
+		err = json.Unmarshal(res.Value, &cfg)
+		if err != nil {
+			err = q.Close()
+			if err != nil {
+				return err
+			}
+			return err
+		}
+
+		err = r.cacheLens(txn, cfg)
+		if err != nil {
+			err = q.Close()
+			if err != nil {
+				return err
+			}
+			return err
+		}
+	}
+
+	err = q.Close()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *lensRegistry) MigrateUp(
+	src enumerable.Enumerable[LensDoc],
+	schemaVersionID string,
+) (enumerable.Enumerable[LensDoc], error) {
+	lensLocker, ok := r.lensWarehouse[schemaVersionID]
+	if !ok {
+		// If there are no migrations for this schema version, just return the given source.
+		return src, nil
+	}
+
+	lens, err := lensLocker.borrow()
+	if err != nil {
+		return nil, err
+	}
+
+	lens.SetSource(src)
+
+	return lens, nil
+}
+
+func (*lensRegistry) MigrateDown(
+	src enumerable.Enumerable[LensDoc],
+	schemaVersionID string,
+) (enumerable.Enumerable[LensDoc], error) {
+	// todo: https://github.com/sourcenetwork/defradb/issues/1591
+	return src, nil
+}
+
+func (r *lensRegistry) Config() []client.LensConfig {
+	result := []client.LensConfig{}
+	for _, cfg := range r.lensConfigs {
+		result = append(result, cfg)
+	}
+	return result
+}
+
+// lensLocker provides a pool-like mechanic for caching a limited number of wasm lens modules in
+// a thread safe fashion.
+//
+// Instanstiating a lens module is pretty expensive as it has to spin up the wasm runtime environment
+// so we need to limit how frequently we do this.
+type lensLocker struct {
+	// The config used to create the lenses within this locker.
+	cfg client.LensConfig
+
+	// Using a buffered channel provides an easy way to manage a finite
+	// number of lenses.
+	//
+	// We wish to limit this as creating lenses is expensive, and we do not want
+	// to be dynamically resizing this collection and spinning up new lens instances
+	// in user time, or holding on to large numbers of them.
+	safes chan *lensPipe
```

Comment:
Yeah, `warehouse`=>`locker`=>`safe`. Naming isn't perfect here, but perfect is an impossibility and I see the names as good enough.  Especially given that most of the week's worth of non-question-comments on this 4000 line PR have been pretty nitpicky, covering almost entirely naming/documentation-tweaks/etc and I'd like to merge this and actually get back to improving the code.

Reply:
> Naming and documenting is quite important 

Of course it is. There is a balance to be had though, if we spend all year haggling over the names of private variables then Source will run out of money and we'll have to find new employment.

Reply:
The nitpick/suggestion comments are specifically trying to ensure future maintenance, always operating under the assumption that devs will forget their exact state of mind they had while writing some specific code. Which is why it's important to use common phrasing, ideas, and metaphors that is either idiomatic to the language or to the
Team. How will the next 10 engineers interpret things?

All of programming is just a game of metaphors and analogies we use to convey meaning.

I'd rather spend an extra week on a PR highlighting and discussing potential futures maintenance issues, rather than find ourselves in the future battling more and more refactors.

It's a situation where an ounce of prevention is worth a pound of cure. So to trivialize these kinds of discussion which may feel pedantic, as "chit chat" is a little disingenuous.

Reply:
> actually get back to improving the code.

This is improving the code

Reply:
Any preferred alternatives?

Reply:
how about `schemaPipePools=>pipePool=>pipe`?

Reply:
I do think `pool(s)` is fairly common/understood in this context as they often reference "pools of resources" which can be seen at every level of software and hardware design. Even go has a [Pool](https://pkg.go.dev/sync#Pool) which itself is defined as:

*"An appropriate use of a Pool is to manage a group of temporary items silently shared among and potentially reused by concurrent independent clients of a package."*

Reply:
> I do think pool(s) is fairly common/understood 

I am not disputing that.  But that doesn't mean we should expand the use for that reason alone. Or ignore the `pipe`-`pool` issue I mentioned.

Reply:
I feel its a better analogy than locker/safe/warehouse combo. I see little issue with pool/pipe clashing, and again fairly ubiquitous in Go.

Reply:
Renamed